### PR TITLE
src: New Message Fields Support

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

~~Bring in support for https://github.com/discordapp/discord-api-docs/pull/1030 when it gets merged~~

Brings discord.js up to date with https://github.com/discordapp/discord-api-docs/pull/1051

Kept as draft till the mentioned PR gets merged

**Status**
- [ ] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [ ] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by NightYoshi370** [comment]
The PR linked in the main post has been closed. However, this comes to replace it: https://github.com/discordapp/discord-api-docs/pull/1051

File: src/structures/MessageMentions.js [hunk 1]
```diff
@@ -86,6 +87,29 @@ class MessageMentions {
...
+     * @typedef {Object} CrosspostedChannel
+     * @property {string} channelID ID of the mentioned channel
+     * @property {string} guildID ID of the guild that has the channel
+     * @property {string} type Type of the channel
+     * @property {string} name The name of the channel
+     */
+
+    /**
+     * An array of crossposted channels
+     * @type {?CrosspostedChannel[]}
```
Comment (SpaceEEC) [inline src/structures/MessageMentions.js]: Shouldn't this be a `?Collection<Snowflake, CrosspostedChannel>`?
Although if available I think we should include the actual channel, so `?Collection<Snowflake, CrosspostedChannel | NewsChannel>`.

File: src/structures/MessageMentions.js [hunk 2]
```diff
@@ -86,6 +87,29 @@ class MessageMentions {
...
+     */
+    this.crosspostedChannels = crosspostedChannels ? crosspostedChannels.map(d => {
```
Comment (SpaceEEC) [inline src/structures/MessageMentions.js]: This is missing typings :eyes:

File: src/structures/MessageMentions.js (hunk 1 above)
Comment (vladfrangu) [reply src/structures/MessageMentions.js]: A collection would be useful, but having the type be `CrosspostedChannel | NewsChannel` would be a bit confusing (as CrosspostedChannel isn't a true "channel" like the NewsChannel)

File: typings/index.d.ts [hunk 3]
```diff
@@ -2596,5 +2615,23 @@ declare module 'discord.js' {
 	type CloseEvent = { wasClean: boolean; code: number; reason: string; target: WebSocket; };
 	type ErrorEvent = { error: any, message: string, type: string, target: WebSocket; };
 
+	enum ChannelType {
```
Comment (SpaceEEC) [inline typings/index.d.ts]: discord.js does not export such an enum.
Maybe it's possible to declare it to be "type-only", so you can not use it as a value.

Comment (vladfrangu) [reply typings/index.d.ts]: I'd love to know why TS exports this automatically, or how I could declare it without it being exported... Maybe moving it outside the `declare module`?

**Comment by iCrawl** [comment]
@SpaceEEC actually requesting a review again please

File: src/structures/Message.js [hunk 4]
```diff
@@ -172,6 +175,30 @@ class Message extends Base {
...
+     * @type {Readonly<MessageFlags>}
+     */
+    this.flags = new MessageFlags(data.flags).freeze();
+
+    /**
+     * Reference data sent in a crossposted message.
+     * @typedef {Object} MessageReference
+     * @property {string} channelID ID of the channel the message was crossposted from
+     * @property {string} guildID ID of the guild the message was crossposted from
+     * @property {string} [messageID] ID of the message that was crossposted
```
Comment (SpaceEEC) [inline src/structures/Message.js]: The property is guaranteed, it just can be null, so shouldn't it be that then?
```suggestion
     * @property {?string} messageID ID of the message that was crossposted
```

File: src/structures/Message.js [hunk 5]
```diff
@@ -214,8 +241,11 @@ class Message extends Base {
       this,
       'mentions' in data ? data.mentions : this.mentions.users,
       'mentions_roles' in data ? data.mentions_roles : this.mentions.roles,
-      'mention_everyone' in data ? data.mention_everyone : this.mentions.everyone
+      'mention_everyone' in data ? data.mention_everyone : this.mentions.everyone,
+      'mention_channels' in data ? data.mention_channels : this.mentions.crosspostedChannels
     );
+
+    this.flags = new MessageFlags('flags' in data ? data.flags : this.flags.bitfield).freeze();
```
Comment (SpaceEEC) [inline src/structures/Message.js]: It looks like Discord does not send `flags` at all if they are 0.
This leads to `SUPPRESS_EMBEDS` not getting cleared if you unsupress a message.
(Although that endpoint is at the moment technically not """officially supported""" by Discord)

File: src/structures/Message.js (hunk 4 above)
Comment (vladfrangu) [reply src/structures/Message.js]: image

which per discord's docs:

image

So, no. The property isn't guaranteed. Well, it is for basically everything current but for the first crossposted messages, it ain't

File: src/structures/Message.js (hunk 5 above)
Comment (vladfrangu) [reply src/structures/Message.js]: Would it be safe to assume that if `'flags' in data` isn't true, we don't have a message field? How do we even test that? Can you suppress the embeds of a crossposted message? If so, I'll change that 👍

Comment (SpaceEEC) [reply src/structures/Message.js]: > Would it be safe to assume that if `'flags' in data` isn't true, we don't have a message field?

What `message` field do you mean?

From what I've seen during a bit of testing it seems like the field is missing if no flags are present.
I haven't seen any cases where there were any flags but the property missing.
>Can you suppress the embeds of a crossposted message?  

Yes, you can. Both flags are present then, unsuppressing them later sends another update with just the crosspost flag. (Behavior is as you'd expect it)

Comment (vladfrangu) [reply src/structures/Message.js]: > What `message` field do you mean?
I meant message flags. Close enough 😅 


> Yes, you can. Both flags are present then, unsuppressing them later sends another update with just the crosspost flag. (Behavior is as you'd expect it)

Perfect! Then, we default to 0 if 'flags' isn't in data when we patch?

File: src/structures/Message.js (hunk 4 above)
Comment (vladfrangu) [reply src/structures/Message.js]: I just noticed guild_id is also... optional. Can you make your suggestion be for the guild one too, I'll commit it!

**Pull request merged**